protocol/common/message_segmentation.go

```diff
@@ -0,0 +1,361 @@
+func (s *MessageSender) handleSegmentationLayerV2(message *v1protocol.StatusMessage) error {
+	logger := s.logger.With(zap.String("site", "handleSegmentationLayerV2")).With(zap.String("hash", types.HexBytes(message.TransportLayer.Hash).String()))
+
+	segmentMessage := &SegmentMessage{
+		SegmentMessage: &protobuf.SegmentMessage{},
+	}
+	err := proto.Unmarshal(message.TransportLayer.Payload, segmentMessage.SegmentMessage)
+	if err != nil {
+		return errors.Wrap(err, "failed to unmarshal SegmentMessage")
+	}
+
+	logger.Debug("handling message segment", zap.String("EntireMessageHash", types.HexBytes(segmentMessage.EntireMessageHash).String()),
+		zap.Uint32("Index", segmentMessage.Index), zap.Uint32("SegmentsCount", segmentMessage.SegmentsCount),
+		zap.Uint32("ParitySegmentIndex", segmentMessage.ParitySegmentIndex), zap.Uint32("ParitySegmentsCount", segmentMessage.ParitySegmentsCount))
+
+	alreadyCompleted, err := s.persistence.IsMessageAlreadyCompleted(segmentMessage.EntireMessageHash)
+	if err != nil {
+		return err
+	}
+	if alreadyCompleted {
+		return ErrMessageSegmentsAlreadyCompleted
+	}
+
+	if !segmentMessage.IsValid() {
+		return ErrMessageSegmentsInvalidCount
+	}
+
+	err = s.persistence.SaveMessageSegment(segmentMessage, message.TransportLayer.SigPubKey, time.Now().Unix())
+	if err != nil {
+		return err
+	}
+
+	segments, err := s.persistence.GetMessageSegments(segmentMessage.EntireMessageHash, message.TransportLayer.SigPubKey)
+	if err != nil {
+		return err
+	}
+
+	if len(segments) == 0 {
+		return errors.New("unexpected state: no segments found after save operation") // This should theoretically never occur.
+	}
+
+	firstSegmentMessage := segments[0]
+	lastSegmentMessage := segments[len(segments)-1]
+
+	// First segment message must not be a parity message.
+	if firstSegmentMessage.IsParityMessage() || len(segments) != int(firstSegmentMessage.SegmentsCount) {
+		return ErrMessageSegmentsIncomplete
+	}
+
+	payloads := make([][]byte, firstSegmentMessage.SegmentsCount+lastSegmentMessage.ParitySegmentsCount)
+	payloadSize := len(firstSegmentMessage.Payload)
+
+	restoreUsingParityData := lastSegmentMessage.IsParityMessage()
+	if restoreUsingParityData {
+		enc, err := reedsolomon.New(int(firstSegmentMessage.SegmentsCount), int(lastSegmentMessage.ParitySegmentsCount))
+		if err != nil {
+			return err
+		}
+
+		var lastNotParitySegmentPayload []byte
+		for _, segment := range segments {
+			if !segment.IsParityMessage() {
+				if segment.Index == firstSegmentMessage.SegmentsCount-1 {
+					// Ensure last segment is aligned to payload size, as it is required by reedsolomon.
+					payloads[segment.Index] = make([]byte, payloadSize)
+					copy(payloads[segment.Index], segment.Payload)
+					lastNotParitySegmentPayload = segment.Payload
+				} else {
+					payloads[segment.Index] = segment.Payload
+				}
+			} else {
+				payloads[firstSegmentMessage.SegmentsCount+segment.ParitySegmentIndex] = segment.Payload
+			}
```

Comment:
```suggestion
			if segment.IsParityMessage() {
				payloads[firstSegmentMessage.SegmentsCount+segment.ParitySegmentIndex] = segment.Payload
				continue
			}
		    if segment.Index != firstSegmentMessage.SegmentsCount-1 {
			    payloads[segment.Index] = segment.Payload
			    continue
		    }
			// Ensure last segment is aligned to payload size, as it is required by reedsolomon.
			payloads[segment.Index] = make([]byte, payloadSize)
			copy(payloads[segment.Index], segment.Payload)
			lastNotParitySegmentPayload = segment.Payload
```

Reply:
I prefer the way it is now, if you don't mind. Even though the indentation level is deeper, the logic reads better to me.